crates/storage/codecs/src/lib.rs

```diff
@@ -82,15 +84,21 @@ where
     T: Compact + Default,
 {
     /// Returns 0 since we won't include it in the `StructFlags`.
-    fn to_compact(self, buf: &mut impl bytes::BufMut) -> usize {
+    fn to_compact<B>(self, buf: &mut B) -> usize
+    where
+        B: bytes::BufMut + AsMut<[u8]>,
+    {
         // TODO: can it be smaller?
         buf.put_u16(self.len() as u16);
 
         for element in self {
-            // TODO: elias fano?
-            let mut inner = Vec::with_capacity(32);
-            buf.put_u16(element.to_compact(&mut inner) as u16);
-            buf.put_slice(&inner);
+            let length_index = buf.as_mut().len();
+            buf.put_slice(&[0, 0]);
```

Comment:
Is this right? Why empty slice?

Reply:
so the current structure is: `<length_u16><compacted_element>`. 
* We only know the size of the element once it has been compacted.
* `BufMut` doesn't allow to move the cursor back, only forward.

Before, we were allocating the compacted element in another temporary buffer and writing the `length` to our `buf`. Afterwards, we were copying from the temporary buffer to our actually `buf`.

This way, we allocate the necessary bytes (2) for writing the `length` on our `buf`, and then write the compacted element to it. We then use `as_mut()` to go back (thus the need for `AsMut<[u8]>`) and replace with the actual length.
This saves the need for allocating that temporary buffer.

Reply:
Sounds good. Can you please add that as a comment?